contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java

```diff
@@ -0,0 +1,626 @@
+import java.util.Comparator;
+import java.util.Formatter;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * HTML page formatter for tracing debug. The page displays information about all active spans and
+ * all sampled spans based on latency and errors.
+ *
+ * <p>It prints a summary table which contains one row for each span name and data about number of
+ * active and sampled spans.
+ *
+ * <p>See {@link TracezHttpHandler} for how to use the formatter with private HTTP server
+ * implementations or how to use with {@link com.sun.net.httpserver.HttpServer}.
+ */
+public final class TracezPageFormatter {
+
+  private static final Tracer tracer = Tracing.getTracer();
+
+  // Color to use for zebra-striping.
+  private static final String ZEBRA_STRIPE_COLOR = "#eee";
+  // The header for span name.
+  private static final String HEADER_SPAN_NAME = "zspanname";
+  // The header for type (running = 0, latency = 1, error = 2) to display.
+  private static final String HEADER_SAMPLES_TYPE = "ztype";
+  private static final int TYPE_RUNNING = 0;
+  private static final int TYPE_LATENCY = 1;
+  private static final int TYPE_ERROR = 2;
+  // The header for sub-type:
+  // * for latency based samples [0, 8] representing the latency buckets, where 0 is the first one;
+  // * for error based samples [0, 15], 0 - means all, otherwise the error code;
+  private static final String HEADER_SAMPLES_SUB_TYPE = "zsubtype";
+
+  // Map from LatencyBucketBoundaries to the human string displayed on the UI for each bucket.
+  private static final Map<LatencyBucketBoundaries, String> LATENCY_BUCKET_BOUNDARIES_STRING_MAP =
+      buildLatencyBucketBoundariesStringMap();
+
+  private final RunningSpanStore runningSpanStore;
+  private final SampledSpanStore sampledSpanStore;
+
+  private TracezPageFormatter(
+      @Nullable RunningSpanStore runningSpanStore, @Nullable SampledSpanStore sampledSpanStore) {
+    this.runningSpanStore = runningSpanStore;
+    this.sampledSpanStore = sampledSpanStore;
+  }
+
+  /**
+   * Constructs a new {@code TracezPageFormatter}.
+   *
+   * @param runningSpanStore the instance of the {@code RunningSpanStore} to be used.
+   * @param sampledSpanStore the instance of the {@code SampledSpanStore} to be used.
+   * @return a new {@code TracezPageFormatter}.
+   */
+  public static TracezPageFormatter create(
+      @Nullable RunningSpanStore runningSpanStore, @Nullable SampledSpanStore sampledSpanStore) {
+    return new TracezPageFormatter(runningSpanStore, sampledSpanStore);
+  }
+
+  /**
+   * Emits the HTML generated page to the {@code outputStream}.
+   *
+   * @param queryMap the query components map.
+   * @param outputStream the output {@code OutputStream}.
+   */
+  public void emitHtml(Map<String, String> queryMap, OutputStream outputStream) {
```

Comment:
Okay.  I wasn't recommending j2html specifically, I just wanted to give an example of what I meant.